samples: drivers: led_pwm on stm32 boards

FRASTM opened the pull request:
Configure with overlay files for the stm32 boards 
- disco_l475_iot1  : green led LD2  and green led LD2
- nucleo_u575zi_q : green led LD2  and blue led LD2
- nucleo_f091rc : green led LD2 
- nucleo_l073rz  : green led LD2 

to pass the samples/drivers/led_pwm

The led(s) used for pwm on the timer channel output should not already be defined as a leds node in the board dts
That's the reason for the overlay removes that `&leds` node (the alias is kept).

Fixes https://github.com/zephyrproject-rtos/zephyr/discussions/53925

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: drivers: led pwm running on 2 pwm leds on disco l475

Add the overlay file to run the samples/drivers/led_pwm
on the disco_l475_iot1 board. Two green leds are used on pa5 and pb14
corresponding on 2 pwm channels of timers2 and timers15.
The leds node has to be disabled to let the pin for the pwm nodes.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: drivers: led pwm running on pwm led of the nucleo boards

Configure the overlay file to run the samples/drivers/led_pwm
on the nucleo_l073rz and nucleo_f091rc board.
The green led is used on pa5 for pwm channels of timers2 channel
The leds node has to be disabled to let the pin for the pwm nodes.
The spi1 is disabled in that sample to avoid conflict on pa5 pin.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

simonguinot commented on samples/drivers/led_pwm/boards/nucleo_f091rc.overlay:
```diff
@@ -12,3 +21,7 @@
 };
+
+&spi1 {
+	status = "okay";
```
In commit message, you say that spi1 is disabled to avoid a conflict. But here it is enabled :)

FRASTM replied:
right, sorry for this copy/paste error

simonguinot commented on samples/drivers/led_pwm/boards/disco_l475_iot1.overlay:
```diff
@@ -0,0 +1,50 @@
+		pinctrl-0 = <&tim15_ch1_pb14>; /* CN1 D10 */
+		pinctrl-names = "default";
+	};
+};
```
Shouldn't you add this nodes into the board dts file, eventually with a disabled status, to avoid a conflict with gpio-led nodes ? I can see it is done like that for nucleo_f091rc. And in this overlay file we would just have to configure the status of the different nodes to switch from gpio-leds to pwm-leds.

simonguinot replied:
As a user I would expect to find all the devices defined in the board dts files. And I would probably miss this overlay files in the led-pwm sample. But obviously it is your call !

simonguinot commented on samples/drivers/led_pwm/boards/nucleo_u575zi_q.overlay:
```diff
@@ -0,0 +1,48 @@
+		pinctrl-names = "default";
+		status = "okay";
+	};
+};
```
Same comments here as for the disco_l475 board.

simonguinot reviewed (changes requested):
Hi @FRASTM,

I find a little bit odd to have this pwm-leds nodes in samples overlay files rather than in DTS board files.

Please see my comments/questions below.

FRASTM commented:
disabling spi1 for the nucleo_f091rc.overlay

FRASTM commented:
During the init phase, if there is a led node, the led_gpio_init() is called during the z_sys_init_run_level() after the pwm_stm32_init. In this case the pin config for PWM (example pa5 alternate function) is overwritten by the led_init pin config.

simonguinot commented:
> During the init phase, if there is a led node, the led_gpio_init() is called during the z_sys_init_run_level() after the pwm_stm32_init. In this case the pin config for PWM (example pa5 alternate function AF2) is overwritten by the led_init pin config.

Yes I understand that you can't have both gpio-leds and pwm-leds nodes enabled at the same time. But maybe you can have both nodes defined in the board dts file with one of them disabled (with a comment explaining it is an alternative LED configuration). Probably you already thought about that... I am only saying it is odd to have the pwm-leds nodes defined in  sample overlays and not in board dts files. A user may not think to look at this place. But if you are OK with that, then I am too :)

erwango reviewed (changes requested):
I'd like to understand why is this conflict happening.
Maybe led_pwm driver is not compatible with STM32 implementation and requires some adaptation.

FRASTM commented:
> I'd like to understand why is this conflict happening. Maybe led_pwm driver is not compatible with STM32 implementation and requires some adaptation.


During the z_sys_init_run_level , init of the pwm comes before init of the led, consequently when pwm_stm32_init is correctly initializing pin (i.e. pa5 ) for pwm output, then the led_gpio_init is overwriting that pin config.
This is more a question of returning from led_gpio_init in case the pin is already used.

simonguinot commented:
> I'd like to understand why is this conflict happening.
> Maybe led_pwm driver is not compatible with STM32 implementation and requires some adaptation.

Hi @erwango,

Apparently you can configure/mux LED pins as GPIO or PWM. So you can use the led-gpio or led-pwm drivers with these LEDs. And the selection is done in dts by disabling nodes.

Now, why do you want to give the priority to the gpio-led driver, knowing that pwm-led is more powerful (it allows to configure brightness and provides some hardware blinking), I don't know....

Also I think these conflicts/options should be described in boards dts files to give users a full picture. They are unlikely to find they can use the pwm-led driver if the nodes are stored/buried in sample overlays.

erwango commented:
I think this is rather an issue in `driver/led`.
Enabling `CONFIG_LED=y` in `samples/drivers/led_pwm` enables `drivers/led_gpio`, while we don't actually this driver for this sample. I think miss a way to distinguish those 2 drivers using Kconfig. This is a side effect of https://github.com/zephyrproject-rtos/zephyr/pull/48827.
I think it should be revisited. Sample should have the ability to select `LED_PWM` and not `LED`.

FRASTM commented:
WA with overlay disabling the leds node (instead of deleting)

simonguinot commented:
> I think this is rather an issue in `driver/led`. Enabling `CONFIG_LED=y` in `samples/drivers/led_pwm` enables `drivers/led_gpio`, while we don't actually this driver for this sample.

All the LED drivers are enabled by default if there is a matching "okay" node in board DT. It is pretty much how all drivers are enabled in Zephyr Kconfigs.

Because the mux configuration is static then you can't have two drivers nodes using the same pins and both set to "okay" in DT. It makes sense, isn't it ?

Now that being said, the LED drivers are only enabled by default from DT. You can still enable/disable Kconfig options as you wish. For example for the led-pwm sample you can add a samples/drivers/led_pwm/boards/$board.conf file for your board and disable the LED drivers you don't want it it.

FRASTM commented:
The samples/drivers/led_pwm/prj.conf does define the `CONFIG_LED=y`  and with CONFIG_LED=n    building fails

simonguinot commented:
> The samples/drivers/led_pwm/prj.conf does define the `CONFIG_LED=y` and with CONFIG_LED=n building fails

Come on... The idea would be to add `CONFIG_LED_GPIO=n` to samples/drivers/led_pwm/boards/$board.conf. Obviously you can't set `CONFIG_LED=n`, this would disable the whole LED subsystem...

Please note that having two nodes sharing a same resource set to both "okay" in a board DT is not correct. Even if you can disable LED drivers from Kconfig, still you should fix your board DT.

In addition, you should prefer the pwm-led driver to led-gpio. As I explained earlier, more features are supported (brightness configuration, hardware blink) and it would be better for users.

simonguinot commented:
> Please note that having two nodes sharing a same resource set to both "okay" in a board DT is not correct.

Well after a second thought, I am no longer sure. Maybe it should be possible :)

Anyway, you have two options to handle the pin conflict between the GPIO and PWM LED devices:

1. from DT, by disabling the device you don't want.
2. from Kconfig, by disabling the driver you don't want.

And the DT nodes for theses devices should all go in the board dts file, not in sample DT overlay.

I think it is OK to automatically enable the drivers for the "okay" device nodes found in DT. There are few cases (like here) where you may need to disable a driver from a configuration file. But it is still better than explicitly enabling all the drivers we need.

erwango commented:
> But it is still better than having to explicitly enable all the drivers we need.

Explicitly enabling the driver we needs is exactly the way it's supposed to be done. See https://docs.zephyrproject.org/latest/build/dts/dt-vs-kconfig.html:
> Use Kconfig to configure software support to build into the final image. Examples include whether to add networking support, **which drivers are needed by the application**, etc.
```

simonguinot commented:
> > But it is still better than having to explicitly enable all the drivers we need.
> 
> Explicitly enabling the driver we needs is exactly the way it's supposed to be done. See https://docs.zephyrproject.org/latest/build/dts/dt-vs-kconfig.html:

Well... here Kconfig still controls software support: you can enable LED or not. But once enabled then the drivers are automatically selected from DT. And you can disable them from Kconfig. So it does match the documentation.

Please note this behavior is not specific to the LED subsystem. It is done this way too in other drivers subsystems.

The current Kconfig/DT configuration mechanism allows to deal with the PWM/GPIO LED conflicts on STM32 boards. So I think you should move forward with this PR.

simonguinot commented:
> > But it is still better than having to explicitly enable all the drivers we need.
> 
> Explicitly enabling the driver we needs is exactly the way it's supposed to be done. See https://docs.zephyrproject.org/latest/build/dts/dt-vs-kconfig.html:
> 
> > Use Kconfig to configure software support to build into the final image. Examples include whether to add networking support, **which drivers are needed by the application**, etc.

Well I think it is a question on interpretation. To me, Kconfig effectively allows to configure the drivers. You can indeed add or remove them as you wish. Just they are *pre-selected* by default if a corresponding "okay" node is found in DT and if the susbsytem is enabled. And this pre-selection behavior is not documented... Honestly I kind of like this "magic" pre-selection mechanism. But I agree it can be a bit disturbing and it should be at least documented.

Anyway, I don't think it is the right place for this discussion. It is a much larger topic that LEDs.

About this PR, @FRASTM and you should go with the current implementation. And if something needs to be fixed, then it can be revisited later along with all the other drivers and samples.

FRASTM committed:
boards: arm: stm32l475 board add nodes for timers15 and pwm leds

Add a new node to the DTS of the disco_l475_iot1 board for timers15
Output pin for the pwm channel1 is PB14 (D10 of arduino CN1)
Add the nodes for pwm leds corresponding to the pwm channels
of the timer 2 and timer 15.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32u575 nucleo add nodes for timers3 & 4 and pwm leds

Add and enable the timers3 and timers4 of the nucleo_u575zi_q.
Both PWM channel 2 of each timer have output pins connected to leds
on the board.
Add the nodes for pwm leds corresponding to the pwm channels
of the timer 3 and timer 4.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: drivers: led pwm running on 2 pwm leds on nucleo_u575

Add the overlay file to run the samples/drivers/led_pwm
on the nucelo_u575zi_q board.
Two green leds are used on pc7 and pb7 corresponding to pwm
channels of timers3 and timers4.
The leds node has to be disabled to let the pin for the pwm nodes.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Rebase on  8512bd18c66699046dffb7da0b02d72301fe8e5b
Definining the timers/pwm nodes directly in the DTS of the boards

simonguinot reviewed (commented):
I still have the same comments...

All the device nodes should be defined in boards DTS files, and not in samples DTS overlays.

And in the led_pwm sample you can either use:

- Board configuration (`boards/$board.conf`) to enable the led_pwm driver and disable led_gpio.
- Or DTS overlay (`boards/$board.overlay`) to enable `"pwm-leds"` compatible nodes and disable `"gpio-leds"`.

We already discussed the former option in previous comments. Did you try it ?

With this PR we have all the LEDs nodes (PWM and GPIO) for nucleo_f091rc and nucleo_l073rz defined in boards DTS. But for disco_l475_iot1 and nucleo_u575zi_q the `"pwm-leds"` nodes are defined in the led_pwm sample. It is not consistent.

I am not requesting this changes because I think they are more of a board matter than a LED one. But I don't approve this PR either.

FRASTM commented:
Rebase on 60a20471b561a0a3a74b377991fe1976c2ea83d7
Put `pwmleds` nodes in the DTS of each board. Keep the overlay of the sample for activation.

simonguinot reviewed (changes requested):
Hi @FRASTM

The changes are looking good to me. So you finally gave up the idea of using Kconfig to enable/disable the LED drivers ?  This is fine for me but I felt like @erwango had some preference with this option.

Now, about your commit messages... they are messy. For example you have two commits with the exactly same title: `samples: drivers: led pwm running on 2 pwm leds of the board`. Which board are you referring to btw ?

Also you have a weird formatting in some places. For example:

```
boards: arm: stm32l475 disco kit add a node for pwmleds and timers15
    
Add a new node to the DTS of the disco_l475_iot1 board
for the timers15.
Add the pwmleds definitions for timers15 and timers2.
```

Your title is way longer than the lines in body. And this lines are cut shorter than in other commits messages.

Please, can you clean-up a little bit ?

FRASTM commented:
rebase on fcd60cf789b1acd1cd29dcd8db2dba86572bdbfa
Rewrite commit messages : keeping on per board change

simonguinot commented:
> rebase on [fcd60cf](https://github.com/zephyrproject-rtos/zephyr/commit/fcd60cf789b1acd1cd29dcd8db2dba86572bdbfa) Rewrite commit messages : keeping on per board change

Hi @FRASTM. It looks like you not only rewrote the commit messages but also moved back the pwm-leds nodes into the sample overlays... So are we back to the version you pushed 5 days ago ?

FRASTM commented:
> > rebase on [fcd60cf](https://github.com/zephyrproject-rtos/zephyr/commit/fcd60cf789b1acd1cd29dcd8db2dba86572bdbfa) Rewrite commit messages : keeping on per board change
> 
> Hi @FRASTM. It looks like you not only rewrote the commit messages but also moved back the pwm-leds nodes into the sample overlays... So are we back to the version you pushed 5 days ago ?

 + Retrieving the right commit content

erwango reviewed (approved):
Approved, but raising an issue on the Kconfig handling in LED subsystem

simonguinot commented:
Hi @FRASTM. It looks good to me. I am waiting a little bit on #54609 before approving this PR.

simonguinot reviewed (approved):
Let's not block this PR any longer.

Mani-Sadhasivam reviewed (approved):
Changes LGTM. But I do have a question. Is there any reason for not removing `leds-gpio` and using `pwn-leds` alone? Only downside I could think of is the lack of `blinky` sample.
